spec/ics-2-consensus-verification/README.md

```diff
@@ -0,0 +1,190 @@
+  r :: AccumulatorRoot)`, where
+    * `p` is the commit proof used by `RootOfTrust.v` to verify
+    * `v` is the new verifier, if needed to be updated
+    * `r` is the new `AccumulatorRoot` which will replace the existing one
+ 
+* `Consensus` is a blockchain consensus algorithm which generates valid `Header`s.
+  Defined as a function `RootOfTrust -> [Transaction] -> Header`
+
+* `Blockchain` is a subset of `(Consensus, RootOfTrust, [Header])`, generated 
+by a `Consensus`.
+
+#### Requirements
+
+Consensus verification requires the following accumulator primitives with datastructures and
+properties as defined in ICS23:
+
+* `AccumulatorRoot`
+
+#### Subprotocols
+
+##### Verifier
+
+Verifiers prove new `Header` generated from a blockchain. It is expected to prove a `Header` 
+efficiently; more efficiently than replaying `Consensus` logic for given parent `Header` and the
+transactions. `Header.p` provides the proof that the verifier can use. Verifiers assume the
+following properties will be satisfied for the `Header`s:
+
+1. `Header`s have no more than one direct child
+ 
+* Satisfied if: deterministic safety
+* Possible violation scenario: validator double signing, deep chain reorg
+
+2. `Header`s have at least one direct child
+
+* Satisfied if: liveness, lightclient verifier continuability
+* Possible violation scenario: synchronized halt, incompatible hard fork
+
+3. `Header`s' accumulator root are valid transition from the parents'
+
+* Satisfied if: decentralized block generation, well implemented state machine
+* Possible violation scenario: invariant break, validator cartel
+
+// XXX: should it be described on the connection spec?
+If a `Header` is submitted to the chain(and optionally passed some challenge period for fraud 
+proof), then it is assured that the packet is finalized so the application logic can process it.
+If a `Header` is proven to violate one of the properties, but still can be verified, the tracking 
+chain should detect and take action on the event to prevent further impact. See (link for the ICS 
+for equivocation and fraud proof) for details.
+
+##### Accumulator Root
+
+`RootOfTrust` contains an accumulator root, which identifies the whole state of the 
+corresponding blockchain at the point of time that the commit is generated. It is expected that 
+the verifying inclusion or exclusion of certain data in the accumulator is done efficient. See 
+ICS23 for the details about the `AccumulatorRoot`s.
+
+##### Consensus 
+
+`Consensus` is a blockchain protocol which actually generates a list of `Header` from the latest
+state and the incoming transactions. While the chains on the network does not directly proving the 
+consensus process, it is expected that the consensus algorithms will generate valid `Header`s.
+
+### Example Implementation
+
+An example blockchain `B` runs on a single operator consensus algorithm, called `Op`. If a 
```

Comment:
This is a great example! Do you think we can use at least two or three-letter variable names? It would make it easier to read. e.g. - `Hdr`